src/main/python/deep_qa/layers/entailment_models/tuple_match.py

```diff
@@ -0,0 +1,180 @@
+                A 0/1 corresponding to disabling/enabling biases in each of the hidden layers
+        """
+
+    def __init__(self, params: Dict[str, Any]):
+        """
+        Construct a new TupleMatch layer.
+
+        Parameters
+        ----------
+
+        """
+        self.input_dim = None
+        self.supports_masking = True
+        # Parameters for the shallow neural network
+        self.num_entailment_features = params.pop('num_entailment_features', 4)
+        self.num_hidden_layers = params.pop('num_hidden_layers', 1)
+        self.hidden_layer_width = params.pop('hidden_layer_width', 4)
+        self.hidden_layer_activation = params.pop('hidden_layer_activation', 'relu')
+        self.use_biases = bool(params.pop('use_bias', 1))
+        # NN layers
+        self.hidden_layers = None
+        self.score_layer = None
+        super(TupleMatch, self).__init__(**params)
+
+    def score_tuple_entailment(self, combined_tuple_input):
+        """
+        Here we take the combined tuple-based entailment input and decide whether it is true or false. The
+        combined input has shape (batch_size, combined_input_dim).  We do not know what
+        combined_input_dim is, as it depends on the combiner.
+        """
+        hidden_input = combined_tuple_input
+        for layer in self.hidden_layers:
+            hidden_input = layer(hidden_input)
+        score_output = self.score_layer(hidden_input)
+        return score_output
+
+
+    def get_output_shape_for(self, input_shapes):
+        #pylint: disable=unused-argument
+        return (1,)
+
+
+    def build(self, input_shape):
+        # Build the dense layer(s)
+        self.hidden_layers = []
+        dense_input_shape = (None, self.num_entailment_features)
+        for i in range(self.num_hidden_layers):
+            curr_dense = Dense(output_dim=self.hidden_layer_width,
+                               activation=self.hidden_layer_activation,
+                               bias=self.use_biases,
+                               name='tuple_match_hidden_layer_%d' % i)
+            curr_dense.build(input_shape=dense_input_shape)
+            dense_input_shape = (None, self.hidden_layer_width)
+            self.hidden_layers.append(curr_dense)
+        score_layer_dense = Dense(output_dim=1,
+                                  activation='sigmoid',
+                                  name='tuple_match_score',
+                                  bias=self.use_biases)
+        score_layer_dense.build(input_shape=dense_input_shape)
+        self.score_layer = score_layer_dense
+
+        # Add the weights from these layers to TupleMatch's trainable weights
+        for hidden in self.hidden_layers + [self.score_layer]:
+            self.trainable_weights.extend(hidden.trainable_weights)
+
+        super(TupleMatch, self).build(input_shape)
+
+    @overrides
+    def compute_mask(self, inputs, input_mask=None):  # pylint: disable=unused-argument
```

Comment:
Yeah, it's an awful decision that keras made.  But we have to follow that decision, or our code will break.  I ran into a crash because of this a while ago.  It's really unfortunate.